postrs/bridge.go

```diff
@@ -0,0 +1,138 @@
+package postrs
+
+// #cgo LDFLAGS: -lpost
+// #include <stdlib.h>
+// #include "prover.h"
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"reflect"
+	"unsafe"
+
+	"github.com/spacemeshos/post/config"
+	"github.com/spacemeshos/post/shared"
+)
+
+// Translate scrypt parameters expressed as N,R,P to Nfactor, Rfactor and Pfactor
+// that are understood by scrypt-jane.
+// Relation:
+// N = 1 << (nfactor + 1)
+// r = 1 << rfactor
+// p = 1 << pfactor
+func translateScryptParams(params config.ScryptParams) C.ScryptParams {
+	return C.ScryptParams{
+		nfactor: C.uint8_t(math.Log2(float64(params.N))) - 1,
+		rfactor: C.uint8_t(math.Log2(float64(params.R))),
+		pfactor: C.uint8_t(math.Log2(float64(params.P))),
+	}
+}
+
+func GenerateProof(datadir string, challenge []byte, cfg config.Config, nonces uint, powScrypt config.ScryptParams) (*shared.Proof, error) {
+	datadirPtr := C.CString(datadir)
+	defer C.free(unsafe.Pointer(datadirPtr))
+
+	challengePtr := C.CBytes(challenge)
+	defer C.free(challengePtr)
+
+	config := C.Config{
+		k1:                C.uint32_t(cfg.K1),
+		k2:                C.uint32_t(cfg.K2),
+		k2_pow_difficulty: C.uint64_t(cfg.K2PowDifficulty),
+		k3_pow_difficulty: C.uint64_t(cfg.K3PowDifficulty),
+		pow_scrypt:        translateScryptParams(powScrypt),
+	}
+
+	cProof := C.generate_proof(
+		datadirPtr,
+		(*C.uchar)(challengePtr),
+		C.size_t(len(challenge)),
+		config,
+		C.size_t(nonces),
+	)
+
+	if cProof == nil {
+		return nil, fmt.Errorf("got nil")
+	}
+	defer C.free_proof(cProof)
+
+	indices := make([]uint8, cProof.indices.len)
+	copy(indices, unsafe.Slice((*uint8)(unsafe.Pointer(cProof.indices.ptr)), cProof.indices.len))
+
+	return &shared.Proof{
+		Nonce:   uint32(cProof.nonce),
+		Indices: indices,
+		K2Pow:   uint64(cProof.k2_pow),
+		K3Pow:   uint64(cProof.k3_pow),
+	}, nil
+}
+
+func VerifyProof(proof *shared.Proof, metadata *shared.ProofMetadata, cfg config.Config, theads uint, powScrypt, labelScrypt config.ScryptParams) error {
+	if proof == nil {
+		return errors.New("proof cannot be nil")
+	}
+	if metadata == nil {
+		return errors.New("metadata cannot be nil")
+	}
+	if len(metadata.NodeId) != 32 {
+		return errors.New("node id length must be 32")
+	}
+	if len(metadata.CommitmentAtxId) != 32 {
+		return errors.New("commitment atx id length must be 32")
+	}
+	if len(metadata.Challenge) != 32 {
+		return errors.New("challenge length must be 32")
+	}
+	if len(proof.Indices) == 0 {
+		return errors.New("proof indices are empty")
+	}
+
+	config := C.Config{
+		k1:                C.uint32_t(cfg.K1),
+		k2:                C.uint32_t(cfg.K2),
+		k3:                C.uint32_t(cfg.K3),
+		k2_pow_difficulty: C.uint64_t(cfg.K2PowDifficulty),
+		k3_pow_difficulty: C.uint64_t(cfg.K3PowDifficulty),
```

Comment:
Why the values from `cfg` and not from `metadata`?

Reply:
Because K3 and pow difficulties are not part of metadata. For K1 and K2 it doesn't make any difference as they must be equal.